internal/guest/runtime/hcsv2/uvm.go

```diff
@@ -421,32 +422,48 @@ func (h *Host) CreateContainer(ctx context.Context, id string, settings *prot.VM
-	// construction time.
-
-	// It may be an error to have a security policy but not expose it to the container as
-	// in that case it can never be checked as correct by a verifier.
+	// Write security policy, signed UVM reference and host AMD certificate to
+	// container's rootfs, so that application and sidecar containers can have
+	// access to it. The security policy is required by containers which need to
+	// extract init-time claims found in the security policy. The directory path
+	// containing the files is exposed via UVM_SECURITY_CONTEXT_DIR env var.
+	// It may be an error to have a security policy but not expose it to the
+	// container as in that case it can never be checked as correct by a verifier.
 	if oci.ParseAnnotationsBool(ctx, settings.OCISpecification.Annotations, annotations.UVMSecurityPolicyEnv, true) {
+		securityContextDir, err := os.MkdirTemp(settings.OCISpecification.Root.Path, securitypolicy.SecurityContextDirTemplate)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create security context directory: %w", err)
+		}
+		// Make sure it's readable
+		if err := os.Chmod(securityContextDir, 0744); err != nil {
+			return nil, fmt.Errorf("failed to chmod security context directory: %w", err)
+		}
+
 		encodedPolicy := h.securityPolicyEnforcer.EncodedSecurityPolicy()
 		if len(encodedPolicy) > 0 {
-			secPolicyEnv := fmt.Sprintf("UVM_SECURITY_POLICY=%s", encodedPolicy)
-			settings.OCISpecification.Process.Env = append(settings.OCISpecification.Process.Env, secPolicyEnv)
+			if _, err := writeFileInDir(securityContextDir, securitypolicy.PolicyFilename,
+				bytes.NewReader([]byte(encodedPolicy))); err != nil {
+				return nil, fmt.Errorf("failed to write security policy: %w", err)
+			}
 		}
 		if len(h.uvmReferenceInfo) > 0 {
-			uvmReferenceInfo := fmt.Sprintf("UVM_REFERENCE_INFO=%s", h.uvmReferenceInfo)
-			settings.OCISpecification.Process.Env = append(settings.OCISpecification.Process.Env, uvmReferenceInfo)
+			if _, err := writeFileInDir(securityContextDir, securitypolicy.ReferenceInfoFilename,
+				bytes.NewReader([]byte(h.uvmReferenceInfo))); err != nil {
+				return nil, fmt.Errorf("failed to write UVM reference info: %w", err)
+			}
 		}
-		if len(settings.OCISpecification.Annotations[annotations.HostAMDCertificate]) > 0 {
-			amdCertEnv := fmt.Sprintf("UVM_HOST_AMD_CERTIFICATE=%s", settings.OCISpecification.Annotations[annotations.HostAMDCertificate])
-			settings.OCISpecification.Process.Env = append(settings.OCISpecification.Process.Env, amdCertEnv)
+		hostAMDCert := settings.OCISpecification.Annotations[annotations.HostAMDCertificate]
+		if len(hostAMDCert) > 0 {
+			if _, err := writeFileInDir(securityContextDir, securitypolicy.HostAMDCertFilename,
+				bytes.NewReader([]byte(hostAMDCert))); err != nil {
+				return nil, fmt.Errorf("failed to write host AMD certificate: %w", err)
+			}
+		}
+
+		if len(encodedPolicy) > 0 || len(h.uvmReferenceInfo) > 0 || len(hostAMDCert) > 0 {
```

Comment:
should we move this check to the start of this block (L433) so that we skip creating the directory and all that if the three items are empty?